### ISSUE TITLE
Config setting `remote-address-header` is parsed but never used

### ISSUE THREAD

**Issue by luksow** [akka/akka#16641]
`remote-address-header` in `akka.http.server` namespace is parsed but then never actually used. This renders `extractClientIP` directive useless unless you put a proxy in front of your server to append proper headers.

**Comment by luksow** [comment]
My initial guess is that it should append header somewhere in `HttpRequestParser` or `HttpServer` but I'm not that familiar with akka-http code (yet!). With little guidance I can PR a fix.

**Comment by rkuhn** [comment]
Thanks for reporting and for offering to try your hand on this. While I cannot immediately suggest where to start on this one (@sirthias?) we’ll definitely welcome and review a PR.

**Comment by sirthias** [comment]
This is an interesting one, demonstrating a fundamental problem that we might not yet have solved properly, namely https://github.com/akka/akka/issues/16168.
What we want is have the 

```scala
def serverFlowToTransport(serverFlow: Flow[HttpRequest, HttpResponse],
  settings: ServerSettings, log: LoggingAdapter): Flow[ByteString, ByteString]
```

method (optionally) add a special request header holding the clients IP address.
Since the flow produced by `serverFlowToTransport` is supposed to be reuseable we need to somehow access the `MaterializedMap` from within the flow (after all the clients IP differs per materialization).

/cc @jrudolph

**Comment by drewhk** [comment]
@jrudolph @sirthias this particular problem can be now solved with the materializedValue accessor in graphs, right? Is some other feature missing?

**Comment by jrudolph** [comment]
For now we will implement it on top of the current one-time flows which should be very simple.

Making the server flow reusable is still a major task that involves a lot of restructuring that is currently not in scope as we have a working solution by reconstructing the flow per connection. Also, conceptually it would mean that the `BidiFlow` would need to be able to take per-materialization parameters (we talked about this somewhere else) which could be done by the `serverFlow` providing a hole as materialized value and combining that hole at (after?) materialization. In any case this seems rather like a workaround like a real solution.

**Comment by jrudolph** [comment]
Fixed in #17658.
